youtube/get-tokens.js

```diff
@@ -0,0 +1,24 @@
+const { google } = require("googleapis")
+
+const authorize = async ({
```

Comment:
I Woudl like it if you can return the `auth` instead of a function that we need to call to get it. i think it would be more clear and clean with this way we will make sure we are using the same auth instance even if we need to update multiple videos

Reply:
I see what you mean. I'm inclined to perform the authentication step in get-tokens.js and export the resulting oauth2Client, but how do we make sure it won't do this every time get-tokens is imported ?

This is what I had in mind :

```javascript
const { google } = require("googleapis")

const authorize = async () => {
  console.log("Logging in...")
  const oauth2Client = new google.auth.OAuth2(
    process.env.CLIENT_ID,
    process.env.CLIENT_SECRET,
    process.env.REDIRECT_URL
  )

  oauth2Client.setCredentials({
    access_token: process.env.ACCESS_TOKEN,
    refresh_token: process.env.REFRESH_TOKEN,
    scope: "https://www.googleapis.com/auth/youtube.force-ssl",
  })
  return oauth2Client
}
const auth = authorize()

module.exports = auth
```

Then in index.js :

```javascript
(async () => {
  const auth = await require("./get-tokens")
  const youtubeUrl = process.argv[2]
```

This however leaves a bad taste in my mouth. The module makes some assumptions that the consumers of the module might not know about, like defining env variables and having to await it during import.

Reply:
In this case, i would recommend creating a function `getAuth` that returns the auth instance and in case we already have an instance we return it ( singleton pattern )

so `getAuth` should look like the following:

```js
const { google } = require("googleapis")
let auth = null;

const getAuth = async () => {
    if(auth!==null){
     auth =  new google.auth.OAuth2(
    process.env.CLIENT_ID,
    process.env.CLIENT_SECRET,
    process.env.REDIRECT_URL
  )

  auth.setCredentials({
    access_token: process.env.ACCESS_TOKEN,
    refresh_token: process.env.REFRESH_TOKEN,
    scope: "https://www.googleapis.com/auth/youtube.force-ssl",
  })
}
return auth 
}
```